### DISCUSSION TITLE
How to fully qualify a table with both its database and schema name?

### DISCUSSION THREAD

**Discussion by matthewgdv** [sqlalchemy/sqlalchemy#10102] [Usage Questions] [answered]
I'd just like to say up-front I have read through this issue, which was similar, but it hasn't helped my use-case:

https://github.com/sqlalchemy/sqlalchemy/discussions/8027

Basically, if I try to set the schema to '{db_name}.{schema_name}' the whole thing gets quoted and treated as a single schema, so the solution recommended there is a no-go.

I'm trying to do some INSERT FROM SELECTS across 2 separate databases that have an identical schema which is kept in sync using alembic migrations (one for dev, the other prd), and I need to be able to copy some data across from prd to dev for testing purposes.

here's a code example:

```python
insert(table_dev).from_select(
    [
        table_dev.c.dt,
        table_dev.c.base_currency,
        table_dev.c.target_currency,
        table_dev.c.rate,
    ],
    select(
        table_prd.c.dt,
        table_prd.c.base_currency,
        table_prd.c.target_currency,
        table_prd.c.rate,
    )
)
```

and I would like to have it compiled to this:

```sql
INSERT INTO
    dev.exchange_rates.daily
(
    dt,
    base_currency,
    target_currency,
    rate
) 
SELECT 
    prd.exchange_rates.daily.dt,
    prd.exchange_rates.daily.base_currency,
    prd.exchange_rates.daily.target_currency,
    prd.exchange_rates.daily.rate
FROM 
    prd.exchange_rates.daily;
```

rather than this:

```sql
INSERT INTO
    exchange_rates.daily
(
    dt,
    base_currency,
    target_currency,
    rate
) 
SELECT 
    exchange_rates.daily.dt,
    exchange_rates.daily.base_currency,
    exchange_rates.daily.target_currency,
    exchange_rates.daily.rate
FROM 
    exchange_rates.daily;
```

So, I've actually managed to solve this in an incredibly ugly and inelegant way that is probably prone to breakage when sqlalchemy updates, by monkey-patching some visitor methods on the `SQLCompiler` and the format_table method of the `IdentifierPreparer`, like so:

```python
# ----------------------------------------------------------------------------------------------------

# monkey-patch IdentifierPreparer.format_table


original_format_table = IdentifierPreparer.format_table


def format_table(self: IdentifierPreparer, table: Table, use_schema: bool = True, name: str = None) -> str:
    ret = original_format_table(self, table=table, use_schema=use_schema, name=name)
    return f'{self.quote(table.bind.url.database)}.{ret}' if table.bind else ret


IdentifierPreparer.format_table = format_table

# ----------------------------------------------------------------------------------------------------

# monkey-patch SQLCompiler.visit_table

original_visit_table = SQLCompiler.visit_table


def visit_table(self: SQLCompiler, table: Table, asfrom: bool = False, iscrud: bool = False, ashint: bool = False, fromhints: bool = None, use_schema: bool = True, from_linter: bool = None, **kwargs) -> str:
    ret = original_visit_table(self=self, table=table, asfrom=asfrom, iscrud=iscrud, ashint=ashint, fromhints=fromhints, use_schema=use_schema, from_linter=from_linter, **kwargs)
    return f'{self.preparer.quote(table.bind.url.database)}.{ret}' if table.bind else ret


SQLCompiler.visit_table = visit_table

# ----------------------------------------------------------------------------------------------------

# monkey-patch SQLCompiler.visit_column

original_visit_column = SQLCompiler.visit_column


def visit_column(self: SQLCompiler, column: Column, add_to_result_map: bool = None, include_table: bool = True, result_map_targets: tuple = (), **kwargs) -> str:
    ret = original_visit_column(self=self, column=column, add_to_result_map=add_to_result_map, include_table=include_table, result_map_targets=result_map_targets, **kwargs)
    return f'{self.preparer.quote(column.table.bind.url.database)}.{ret}' if column.table is not None and column.table.bind else ret


SQLCompiler.visit_column = visit_column

# ----------------------------------------------------------------------------------------------------
```

Surely this can't be the 'correct' way of achieving this, right? Is there some option that can be set somewhere that causes the compiler to always fully qualify table and column names with the database when a bind is available to them?

Or else is there at least a less hacky way of achieving more or less the exact thing I've just done?

Thanks in advance.

**Comment by zzzeek** [comment, marked as answer]
if you are using SQL Server, which is the only database I know of that has those long dotted schema names, this is supported directly as follows:

```py
from sqlalchemy import Column
from sqlalchemy import DateTime
from sqlalchemy import Float
from sqlalchemy import insert
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.dialects import mssql

m = MetaData()

table_dev = Table(
    "daily",
    m,
    Column("dt", DateTime),
    Column("base_currency", String),
    Column("target_currency", String),
    Column("rate", Float),
    schema="dev.exchange_rates",
)
table_prd = Table(
    "daily",
    m,
    Column("dt", DateTime),
    Column("base_currency", String),
    Column("target_currency", String),
    Column("rate", Float),
    schema="prd.exchange_rates",
)

stmt = insert(table_dev).from_select(
    [
        table_dev.c.dt,
        table_dev.c.base_currency,
        table_dev.c.target_currency,
        table_dev.c.rate,
    ],
    select(
        table_prd.c.dt,
        table_prd.c.base_currency,
        table_prd.c.target_currency,
        table_prd.c.rate,
    ),
)

print(stmt.compile(dialect=mssql.dialect()))

```

with output:

```
INSERT INTO dev.exchange_rates.daily (dt, base_currency, target_currency, rate) SELECT prd.exchange_rates.daily.dt, prd.exchange_rates.daily.base_currency, prd.exchange_rates.daily.target_currency, prd.exchange_rates.daily.rate 
FROM prd.exchange_rates.daily
```

multipart schema names for MS SQL Server only are documented at https://docs.sqlalchemy.org/en/20/dialects/mssql.html#multipart-schema-names 


if this is some other non-MSSQL database, an additional step to force-unquoting of the schema identifiers can be taken:

```py
from sqlalchemy import Column
from sqlalchemy import DateTime
from sqlalchemy import Float
from sqlalchemy import insert
from sqlalchemy import MetaData
from sqlalchemy import quoted_name
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import Table

m = MetaData()

table_dev = Table(
    "daily",
    m,
    Column("dt", DateTime),
    Column("base_currency", String),
    Column("target_currency", String),
    Column("rate", Float),
    schema=quoted_name("dev.exchange_rates", quote=False),
)
table_prd = Table(
    "daily",
    m,
    Column("dt", DateTime),
    Column("base_currency", String),
    Column("target_currency", String),
    Column("rate", Float),
    schema=quoted_name("prd.exchange_rates", quote=False),
)

stmt = insert(table_dev).from_select(
    [
        table_dev.c.dt,
        table_dev.c.base_currency,
        table_dev.c.target_currency,
        table_dev.c.rate,
    ],
    select(
        table_prd.c.dt,
        table_prd.c.base_currency,
        table_prd.c.target_currency,
        table_prd.c.rate,
    ),
)

print(stmt)

```

with equivalent output:

```
INSERT INTO dev.exchange_rates.daily (dt, base_currency, target_currency, rate) SELECT prd.exchange_rates.daily.dt, prd.exchange_rates.daily.base_currency, prd.exchange_rates.daily.target_currency, prd.exchange_rates.daily.rate 
FROM prd.exchange_rates.daily

```

**Reply by matthewgdv** [reply]
Ah, I never cease to be amazed by your near-instant and consistently useful replies anytime I've ever posted on this project's repo @zzzeek. Just wanted to say that as a data engineer who spends half my working hours using `SQLAlchemy` your work is hugely appreciated.

For context, I'm working with a Redshift database, which introduced cross-DB queries within a cluster a few years ago: 

https://docs.aws.amazon.com/redshift/latest/dg/cross-database-overview.html

Looks like that multi-part schema special-casing doesn't automatically work for Redshift, but using `quoted_name(..., quote=False)` worked perfectly. Thanks so much for your help :)

**Reply by zzzeek** [reply]
we probably need more flexibility in getting these multi-part schema names generally supported then.

**Reply by CaselIT** [reply]
other than better documenting it, I don't think we can do much. On pg this is legal
```sql
create schema "foo.bar";
create table "foo.bar".baz (a int);
select * from "foo.bar".baz;
```

**Reply by matthewgdv** [reply]
For whatever my outsider's perspective is worth, to me it seems there are several options in order of lowest to greatest effort:

1) update the documentation to include keywords like 'fully-qualified', 'dot-separated database, schema, and table name', etc. to make it easier for people to google this functionality. Maybe also include the word 'Redshift'. This would make it more discoverable via search engines, since I had no idea to search for the keyword 'multipart'.
2) an engine setting that causes the compiler to use fully-qualified identifier names on backends that support them when set to 'True' (then each dialect can signal whether supports it via a class attribute, default `False`). This would only work on Tables with a `bind`, unless...
3) Add a database attribute to the `Table` class, similar to the schema attribute. Just like the schema attribute it could be `None` by default to signal 'default database', but could be explicitly set. This would build on top of option 2.

I realize options 2 and especially 3 represent a lot of work and testing for what would probably be considered little gain, but option 1 might not be so hard to do. Regardless, given this discussion is now online and can be found by others in the same boat, it's probably not too important to take any action at all.

I wonder actually, is there a backlog for pipe-dream 'nice-to-haves'? Something that option 2 or 3 could be added to? If so, that would be an possibility, since a member of the community with this use-case might stumble upon it eventually and contribute a PR, with confidence that the change would be accepted if done well.

**Comment by zzzeek** [comment]
> other than better documenting it, I don't think we can do much. On pg this is legal
> 
> ```sql
> create schema "foo.bar";
> create table "foo.bar".baz (a int);
> select * from "foo.bar".baz;
> ```

right, I was thinking more something like `quoted_name()` that delivers the SQL Server behavior:

    schema=dot_separated_name("foo.bar")

or 

    schema=dot_separated_name("foo", "bar")

**Reply by CaselIT** [reply]
ok understood. It could work

**Reply by tboddyspargo** [reply]
FWIW - I would personally get value from a solution that could gracefully quote fully-qualified identifier parts using the dialect's quote character.

It looks to me like a fair number of database systems support these kinds of fully qualified names, but some seem to expect (and most, if not all, support) that each dot-separated part have their own quotes.

- [Postgres](https://www.postgresql.org/docs/8.1/ddl-schemas.html#DDL-SCHEMAS-PATH:~:text=Actually%2C%20the%20even,be%20used%20too%2C)
- [BigQuery](https://github.com/googleapis/python-bigquery-sqlalchemy?tab=readme-ov-file#table-names)
- [Snowflake](https://github.com/snowflakedb/snowflake-sqlalchemy/pull/48/files)
- [Redshift](https://docs.aws.amazon.com/redshift/latest/dg/cross-database-overview.html)
- [Presto/Trino](https://github.com/trinodb/trino-python-client?tab=readme-ov-file#extra-credentials)

The [Snowflake approach ](https://github.com/snowflakedb/snowflake-sqlalchemy/pull/48/files) seemed like interesting work-around to extract the different parts of the schema and quote them independently. Having the ability to gracefully handle schemas that are not already split apart is helpful. Is their approach something that you think has merit as something for core `sqlalchemy`? Or perhaps a recommended approach for dialect developers?

**Reply by CaselIT** [reply]
You mean other than the current solution of using either no quote or quote depending on what you want to do?

**Reply by tboddyspargo** [reply]
Yes. It was my impression that, if not specified, quoting would be applied automatically depending on the casing/characters of the identifier in question.

Since my application will be given a raw, unquoted table name and schema name (with or without additional database identifiers), it would be helpful if the dialect applied quoting (or not) based on the characters in each part of the name. However I’m guessing that dialects may not have convenient hooks for doing that part-by-part for a multipart schema or fully qualified table name.

**Reply by zzzeek** [reply]
> FWIW - I would personally get value from a solution that could gracefully quote fully-qualified identifier parts using the dialect's quote character.

this is supported and these dialects can opt into this behavior, I would post issues with those dialects for them to do so.  The included MSSQL dialect with SQLAlchemy for example supports this, see https://docs.sqlalchemy.org/en/20/dialects/mssql.html#multipart-schema-names .  Those dialects can work from the MSSQL approach to implement in the way that they see fit.  the most critical aspect is making sure there's still a way to have an actual dot in an identifier that's quoted.  The SQL Server dialect uses the brackets to achieve this but for other backends the specfic quote character would likely be more appropriate.

**Reply by tboddyspargo** [reply]
> I would post issues with those dialects for them to do so.

Will do! I'll take a closer look at the mssql implementation to fully appreciate how this is achieved.


> the most critical aspect is making sure there's still a way to have an actual dot in an identifier that's quoted. The SQL Server dialect uses the brackets to achieve this but for other backends the specfic quote character would likely be more appropriate.


I think I misunderstood how to use it appropriately, thank you!

For example, this wasing providing my desired result:
```python
sa.Table("Db_Name.schema Name.test_table", sa.MetaData())
# Compiles to: [Db_Name.schema Name.table_name]
```

But this did:
```python
sa.Table("test_table", sa.MetaData(), schema="Db_Name.schema Name")
# Compiles to: [Db_Name].[schema Name].test_table
```


Is there an approach you'd recommend for parsing the schema out of a fully qualified table name?

**Reply by CaselIT** [reply]
a table name in sqlalchemy should not include the schema name.
You can probably have that by using `quoted_name(..., quote=False)` but you should use schema to have the schema portion of the table

**Reply by zzzeek** [reply]
the SQL Server docs I linked have all of this.   the behavior is entirely local to the "schema" parameter.   all the dot-separated tokens go into that parameter, tablename remains the single name of the table itself.   think of it as the filename, and the schema is the path, if that helps

**Reply by tboddyspargo** [reply]
Understood. Thank you again, and apologies for having you point out what should have probably been clear from a more careful reading of the docs!
